networkx/utils/decorators.py

```diff
@@ -225,3 +226,59 @@ def _open_file(func, *args, **kwargs):
         return result
 
     return _open_file
+
+
+def nodes_or_number(which_args):
+    """Decorator to allow number of nodes or container of nodes.
+
+    Parameters
+    ----------
+    which_args : int or sequence of ints
+        Location of the node arguments in args. Even if the argument is a
+        named positional argument (with a default value), you must specify its
+        index as a positional argument.
+        If more than one node argument is allowed, can be a list of locations.
+
+    Returns
+    -------
+    _nodes_or_numbers : function
+        Function which replaces int args with ranges.
+
+    Examples
+    --------
+    Decorate functions like this::
+
+       @nodes_or_number(0)
+       def empty_graph(nodes):
+           pass
+
+       @nodes_or_number([0,1])
+       def grid_2d_graph(m1, m2, periodic=False):
+           pass
+
+       @nodes_or_number(1)
+       def full_rary_tree(r, n)
+           # r is a number. n can be a number of a list of nodes
+           pass
+    """
+    @decorator
+    def _nodes_or_number(f, *args, **kw):
+        # form tuple of arg positions to be converted.
+        try:
+            iter_wa = iter(which_args)
+        except TypeError:
+            iter_wa = (which_args,)
+        # change each argument in turn
+        new_args = list(args)
+        for i in iter_wa:
+            n = args[i]
+            try:
+                nodes = list(range(n))
+                if n < 0:
+                    msg = "Negative number of nodes not valid:%i"%n
```

Comment:
I've read some of the ```.format``` vs ```%``` flames but might not understand completely. My impression is that ```%``` has some limitations if you don't know the type of the variable, but that the interface is cleaner and prefered by many. ```.format``` has fewer limitations but is not as readable by many. My overall impression was to use ```%``` when you know the type and ```.format``` when you need it to be robust. Here we know ```n``` is an ```int``` (it works in ```range```) so I chose ```%```.  What's your perspective?

I should change  the spacing in any case. :}